Proposal Release Updates (#842)
kubeflow/community: proposals/841-release-process/README.md

## Motivation

Current challenges based on Kubeflow release 1.10:
* Challenge to get component’s release progress during release meetings.
    - Blockers or challenges on the Release are identified later.
    -  Incomplete/Inaccurate or null information on some components about progress/blockers.
* The Release Manager role challenges:
- Only one person is responsible for the Release and decisions.
- Time dedication and commitment are high and for an extended period (6 months), which makes it hard to find candidates for Release.
- The diverse skills required for the role make it even harder for candidates to join.
* Release Documentation
- Working groups need help on building Technical Documentation for Releases.
- Working on technical documentation during the Release, before the feature freeze, can expedite the release process.
* Feature Freeze vs Building Release Candidate
- Teams required more time to finish the Release and later cherry-pick and work on building the Release.
* Release timeframe 
- The Release is now 27 weeks. There are 10 weeks between feature freeze and Release. 
- Release dates and phases are hard to predict in the current process.

### Goals

> * Reduce time to market of Kubeflow Release

Reviewer: ```suggestion
* Reduce the process and overhead and human hours needed
* Move away from the Waterfall model to an agile model that relies on automated tests instead of human labor.
* Reduce time to market of Kubeflow Release
```

* Reduce meeting time and structure to give back time to the Release Team.

> * Focus on contributions from technical documents and source code by reducing meetings and promoting async communication.

Reviewer: ```suggestion
* Focus on contributions from technical documents and source code by reducing meetings and promoting asynchronous communication.
```

* Promote clarity on dates and phases so distributions, working groups, and the whole community can plan accordingly.
* Improve release cycle by moving technical decisions affecting the release to the Release Management team including Kubeflow platform

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

### Release Timelines
KubeFlow Release x.x is moved to Quarterly Release = 12 weeks

![alt text](release.png)

* Notes: Release dates will not be changed unless critical changes are needed.

> Week 9 is code freeze. However, the working groups can use that week to finish any outstanding feature critical for the release.

Reviewer: ```suggestion
```
We want to get away from the waterfall modell.  Our automatic testing is the agile solution and successor to code freeze

* KubeFlow Release minor versions x.x.xx
** Release can be 60-90 days.

### Release Meetings and Communication
* Create a new Release Management team on Slack for asynchronous communication within the team
* Bi-weekly meetings during Software Development (Week 1-Week 8). Weekly meetings between week 9-12 as needed.
* Meetings will focus on roadmap status, including blockers, feature discussions, the overall status of the release, and the help needed.

### Release Documentation

* Build a team of technical writers to be embedded in each working group for the release time frame. 
* Provide roles within the community to boost participation and bring new members into this group.

> * Technical Doc Lead will lead this team; technical documentation can include a blog and slides to announce the Release.

Reviewer: ```suggestion
* Technical Documentation Lead will lead this team; technical documentation can include a blog and slides to announce the Release.
```

### Release Management Team: 
The Release Management Team is composed by: Release Manager, WG Leads ( in some cases Liasons), Technical Doc Lead.
Responsabilities:

- WG leads from each component (ex., Pipelines) must approve a synchronized PR on the Kubeflow Platform/Manifests when related to that component 

> - WG leads from each component (ex., Pipelines) must approve a PR when updates regarding each component on the Kubeflow Platform/Manifests.

Reviewer: why do you want to EXclude pipelines?

> - WG leads from each component (ex., Pipelines) must approve a synchronized PR on the Kubeflow Platform/Manifests when related to that component 
> - WG leads from each component (ex., Pipelines) must approve a PR when updates regarding each component on the Kubeflow Platform/Manifests.

Reviewer: ```suggestion
```

It should be an automatic continuous process. We rely on automatic testing, not manual approvals.

Reviewer: And they are. There are github notifications and this is how we did it over the last years.

- Release team liaisons will be responsible not for the communication but for contributing to the release with documentation, source code, PRs review, etc, according to their skills and motivation.

### Release Manager Responsibilities: 
- Responsible for the overall Kubeflow Release Process 
- Promote best practices for the release and software development process.
- Manage the communication between the teams to understand current release status and potential blockers
- Manage the communication with the community about the status of the Release or any help/blockers needed.
- Synchronize and cut the Release on Kubeflow Platform/Manifests.
- Approve & review the blog and slides announcing the Release.

> ### Implementation -todo update release team
> * Week 0 -  (Release and Roadmap discussions) WG Leads and Release Manager meet to discuss the roadmap planned for the Release
> * Week 2 - (Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting)
> * Week 4 - (Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting)
> * Week 6 - (Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting)
> * Week 8 - (Last week of Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting). Exceptional cases: Discuss any feature that needed to be completed during the feature freeze. 
> * Week 9 - (Feature Freeze and Prep for Release) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting). Discuss items required to prepare for Release (including documentation).
> * Week 10 - (Community and Distribution testing starts) -Sync manifests happen at least the day before. Include Distribution liasions and release team to confirm release testing starts and bug fixing follows.
> * Week 11 - (Bug Fixing) - Discuss any potential blockers for the Release.
> * Week 12 - (Release) - Items to discuss: bug fixes required, release cut, documentation needed.
> - Release Manager synchronizes and cuts the release on Kubeflow Platform/Manifests.
> - WG leads from each component (ex., Pipelines) must approve a synchronized PR on the Kubeflow Platform/Manifests when related to that component 
> - WG leads from each component (ex., Pipelines) must approve a PR when updates regarding each component on the Kubeflow Platform/Manifests.
> - The Release Manager is responsible for approving/reviewing the blog and slides announcing the release.

Reviewer: ```suggestion
### Implementation -todo update release team
* Week 0 -  (Release and Roadmap discussions) WG Leads and Release Manager meet to discuss the roadmap planned for the Release
* Week 2 - (Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting)
* Week 4 - (Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting)
* Week 6 - (Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting)
* Week 8 - (Last week of Software development Phase) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting)
* Week 9 - (Feature Freeze and Prep for Release) WG Leads/Liaisons meet to discuss any release challenges, release changes, and help needed from the community (to communicate on Kubeflow Community Meeting). Discuss items required to prepare for Release (including documentation).
* Week 10 - (Community and Distribution testing starts) Users and distributions run their automated test suites and try to stay away from manual labor-intensive tests.
* Week 11 - (Bug Fixing) - Discuss any potential blockers for the Release.
* Week 12 - (Release) - Items to discuss: bug fixes required, release cut, documentation needed.
- Manifests WG leads synchronizes and cut the release on Kubeflow Platform/Manifests.
- The Release Manager is responsible for approving/reviewing the blog and slides announcing the release.
```

* Focus on contributions from technical documents and source code by reducing meetings and promoting asynchronous communication.
* Promote clarity on dates and phases so distributions, working groups, and the whole community can plan accordingly.
* Improve release cycle by moving technical decisions affecting the release to the Release Management team including Kubeflow platform

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

### Release Timelines
KubeFlow Release x.x is moved to Quarterly Release = 12 weeks

![alt text](release.png)

* Notes: Release dates will not be changed unless critical changes are needed.

* KubeFlow Release minor versions x.x.xx
** Release can be 60-90 days.

### Release Meetings and Communication
* Create a new Release Management team on Slack for asynchronous communication within the team
* Bi-weekly meetings during Software Development (Week 1-Week 8). Weekly meetings between week 9-12 as needed.
* Meetings will focus on roadmap status, including blockers, feature discussions, the overall status of the release, and the help needed.

### Release Documentation

> * Build a team of technical writers to be embedded in each working group for the release time frame. 

Reviewer 2: I think this is fine but we should also try to generate release notes from git commits using semantic commits and GenAI to summarize it.
